lib/spack/spack/binary_distribution.py

```diff
@@ -856,32 +864,87 @@ def _read_specs_and_push_index(file_list, cache_prefix, db, db_root_dir):
-        return
-    except Exception as err:
-        # If we got some kind of S3 (access denied or other connection
-        # error), the first non boto-specific class in the exception
-        # hierarchy is Exception.  Just print a warning and return
-        msg = "Encountered problem listing packages at {0}: {1}".format(cache_prefix, err)
-        tty.warn(msg)
-        return
+    file_list = None
+    tmpspecsdir = None
+    aws = None
+    read_method = None
+
+    def file_read_method(file_path):
+        with open(file_path) as fd:
+            return fd.read()
+
+    def url_read_method(url):
+        contents = None
+        try:
+            _, _, spec_file = web_util.read_from_url(url)
+            contents = codecs.getreader("utf-8")(spec_file).read()
+        except (URLError, web_util.SpackWebError) as url_err:
+            tty.error("Error reading specfile: {0}".format(url))
+            tty.error(url_err)
+        return contents
+
+    if cache_prefix.startswith("s3"):
+        aws = which("aws")
+
+    if aws:
+        tmpspecsdir = tempfile.mkdtemp()
+        sync_command_args = [
+            "s3",
+            "sync",
+            "--exclude",
+            "*",
+            "--include",
+            "*.spec.json.sig",
+            "--include",
+            "*.spec.json",
+            "--include",
+            "*.spec.yaml",
+            cache_prefix,
+            tmpspecsdir,
+        ]
+
+        try:
+            tty.msg(
+                "Using aws s3 sync to download specs from {0} to {1}".format(
+                    cache_prefix, tmpspecsdir
+                )
+            )
+            aws(*sync_command_args, output=os.devnull, error=os.devnull)
+            file_list = fsys.find(tmpspecsdir, ["*.spec.json.sig", "*.spec.json", "*.spec.yaml"])
+            read_method = file_read_method
+        except Exception:
+            tty.msg("Failed to use aws s3 sync to retrieve specs, falling back to parallel fetch")
+            shutil.rmtree(tmpspecsdir)
+            tmpspecsdir = None
+
+    if not file_list:
+        read_method = url_read_method
+        try:
+            file_list = (
+                url_util.join(cache_prefix, entry)
+                for entry in web_util.list_url(cache_prefix)
+                if entry.endswith(".yaml")
+                or entry.endswith("spec.json")
+                or entry.endswith("spec.json.sig")
+            )
+        except KeyError as inst:
+            msg = "No packages at {0}: {1}".format(cache_prefix, inst)
+            tty.warn(msg)
+            return
+        except Exception as err:
+            # If we got some kind of S3 (access denied or other connection
+            # error), the first non boto-specific class in the exception
+            # hierarchy is Exception.  Just print a warning and return
+            msg = "Encountered problem listing packages at {0}: {1}".format(cache_prefix, err)
+            tty.warn(msg)
+            return
```

Comment:
Can we extract different functions here, and have code like:
```console
file_list, read_fn = spec_files_from_cache(cache_prefix)
```
That would make clear that the important result of the first 50 lines of computation is the list of `spec.json.sig` files and the "read function". Also, it would be good to have a single function that retrieves the list using `aws` cli, and another separate function implementing the fallback. 

In this way `spec_files_from_cache` would just be responsible to decide when to call the other two functions.